### PR TITLE
glibc: 2.32-35 -> 2.33-42

### PR DESCRIPTION
**Pull request by lovesegfault**
###### Motivation for this change
https://sourceware.org/pipermail/libc-alpha/2021-February/122207.html
https://hydra.nixos.org/jobset/nixpkgs/glibc-2.33#tabs-evaluations

Todo:
* [x] base onto staging
* [ ] rebuild bootstrap files & upload to tarballs.nixos.org

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by TredwellGit** [comment]
This and https://github.com/NixOS/nixpkgs/pull/110570 both fix CVE-2021-3326.

**Comment by Ma27** [comment]
We'll probably need a staging jobset for this: the last glibc bumps (#95089, #79391, #66528) had a noticeable fallout.

**Comment by lovesegfault** [comment]
> We'll probably need a staging jobset for this: the last glibc bumps (#95089, #79391, #66528) had a noticeable fallout.

Bootstrapping seems broken with this currently.

```
error: --- Error --- nix-daemon
builder for '/nix/store/qipkn62khbv846i0hw39bnkabvl5941l-gcc-10.2.0.drv' failed with exit code 2; last 10 log lines:
  /build/build/./gcc/xgcc -B/build/build/./gcc/ -O2 -I/nix/store/aifxjblpg34q1h1pxc4z5y7lckyas85k-glibc-2.33-0-dev/include -B/nix/store/jgqnaxnyynci109rhxihys0sx6r1b9hy-glibc-2.33-0/lib/ -idirafter /nix/store/aifxjblpg34q1h1pxc4z5y7lckyas85k-glibc-2.33-0-dev/include -idirafter /nix/store/p4s4jf7aq6v6z9iazll1aiqwb34aqxq9-bootstrap-tools/lib/gcc/x86_64-unknown-linux-gnu/8.3.0/include-fixed -Wl,-rpath,/nix/store/50l7bhzcb3v7w2g9xck922f74gjcw93i-gcc-10.2.0-lib/lib -Wl,-L/nix/store/jgqnaxnyynci109rhxihys0sx6r1b9hy-glibc-2.33-0/lib -Wl,-rpath -Wl,/nix/store/jgqnaxnyynci109rhxihys0sx6r1b9hy-glibc-2.33-0/lib -Wl,-dynamic-linker=/nix/store/jgqnaxnyynci109rhxihys0sx6r1b9hy-glibc-2.33-0/lib/ld-linux-x86-64.so.2   -fno-checking -O2  -O2 -I/nix/store/aifxjblpg34q1h1pxc4z5y7lckyas85k-glibc-2.33-0-dev/include -B/nix/store/jgqnaxnyynci109rhxihys0sx6r1b9hy-glibc-2.33-0/lib/ -idirafter /nix/store/aifxjblpg34q1h1pxc4z5y7lckyas85k-glibc-2.33-0-dev/include -idirafter /nix/store/p4s4jf7aq6v6z9iazll1aiqwb34aqxq9-bootstrap-tools/lib/gcc/x86_64-unknown-linux-gnu/8.3.0/include-fixed -Wl,-rpath,/nix/store/50l7bhzcb3v7w2g9xck922f74gjcw93i-gcc-10.2.0-lib/lib -Wl,-L/nix/store/jgqnaxnyynci109rhxihys0sx6r1b9hy-glibc-2.33-0/lib -Wl,-rpath -Wl,/nix/store/jgqnaxnyynci109rhxihys0sx6r1b9hy-glibc-2.33-0/lib -Wl,-dynamic-linker=/nix/store/jgqnaxnyynci109rhxihys0sx6r1b9hy-glibc-2.33-0/lib/ld-linux-x86-64.so.2 -DIN_GCC    -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wno-format -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include   -fpic -mlong-double-80 -DUSE_ELF_SYMVER  -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector  -shared -nodefaultlibs -Wl,--soname=libgcc_s.so.1 -Wl,--version-script=libgcc.map -o ./libgcc_s.so.1.tmp -O2 -I/nix/store/aifxjblpg34q1h1pxc4z5y7lckyas85k-glibc-2.33-0-dev/include -B/nix/store/jgqnaxnyynci109rhxihys0sx6r1b9hy-glibc-2.33-0/lib/ -idirafter /nix/store/aifxjblpg34q1h1pxc4z5y7lckyas85k-glibc-2.33-0-dev/include -idirafter /nix/store/p4s4jf7aq6v6z9iazll1aiqwb34aqxq9-bootstrap-tools/lib/gcc/x86_64-unknown-linux-gnu/8.3.0/include-fixed -Wl,-rpath,/nix/store/50l7bhzcb3v7w2g9xck922f74gjcw93i-gcc-10.2.0-lib/lib -Wl,-L/nix/store/jgqnaxnyynci109rhxihys0sx6r1b9hy-glibc-2.33-0/lib -Wl,-rpath -Wl,/nix/store/jgqnaxnyynci109rhxihys0sx6r1b9hy-glibc-2.33-0/lib -Wl,-dynamic-linker=/nix/store/jgqnaxnyynci109rhxihys0sx6r1b9hy-glibc-2.33-0/lib/ld-linux-x86-64.so.2 -B./ _muldi3_s.o _negdi2_s.o _lshrdi3_s.o _ashldi3_s.o _ashrdi3_s.o _cmpdi2_s.o _ucmpdi2_s.o _clear_cache_s.o _trampoline_s.o __main_s.o _absvsi2_s.o _absvdi2_s.o _addvsi3_s.o _addvdi3_s.o _subvsi3_s.o _subvdi3_s.o _mulvsi3_s.o _mulvdi3_s.o _negvsi2_s.o _negvdi2_s.o _ctors_s.o _ffssi2_s.o _ffsdi2_s.o _clz_s.o _clzsi2_s.o _clzdi2_s.o _ctzsi2_s.o _ctzdi2_s.o _popcount_tab_s.o _popcountsi2_s.o _popcountdi2_s.o _paritysi2_s.o _paritydi2_s.o _powisf2_s.o _powidf2_s.o _powixf2_s.o _mulhc3_s.o _mulsc3_s.o _muldc3_s.o _mulxc3_s.o _divhc3_s.o _divsc3_s.o _divdc3_s.o _divxc3_s.o _bswapsi2_s.o _bswapdi2_s.o _clrsbsi2_s.o _clrsbdi2_s.o _fixunssfsi_s.o _fixunsdfsi_s.o _fixunsxfsi_s.o _fixsfdi_s.o _fixdfdi_s.o _fixxfdi_s.o _fixunssfdi_s.o _fixunsdfdi_s.o _fixunsxfdi_s.o _floatdisf_s.o _floatdidf_s.o _floatdixf_s.o _floatundisf_s.o _floatundidf_s.o _floatundixf_s.o _divdi3_s.o _moddi3_s.o _divmoddi4_s.o _udivdi3_s.o _umoddi3_s.o _udivmoddi4_s.o _udiv_w_sdiv_s.o cpuinfo_s.o sfp-exceptions_s.o addtf3_s.o divtf3_s.o multf3_s.o negtf2_s.o subtf3_s.o unordtf2_s.o fixtfsi_s.o fixunstfsi_s.o floatsitf_s.o floatunsitf_s.o fixtfdi_s.o fixunstfdi_s.o floatditf_s.o floatunditf_s.o fixtfti_s.o fixunstfti_s.o floattitf_s.o floatuntitf_s.o extendsftf2_s.o extenddftf2_s.o extendxftf2_s.o trunctfsf2_s.o trunctfdf2_s.o trunctfxf2_s.o getf2_s.o letf2_s.o eqtf2_s.o _divtc3_s.o _multc3_s.o _powitf2_s.o enable-execute-stack_s.o unwind-dw2_s.o unwind-dw2-fde-dip_s.o unwind-sjlj_s.o unwind-c_s.o emutls_s.o libgcc.a -lc && rm -f ./libgcc_s.so && if [ -f ./libgcc_s.so.1 ]; then mv -f ./libgcc_s.so.1 ./libgcc_s.so.1.backup; else true; fi && mv ./libgcc_s.so.1.tmp ./libgcc_s.so.1 && (echo "/* GNU ld script"; echo "   Use the shared library, but some functions are only in"; echo "   the static library.  */"; echo "GROUP ( libgcc_s.so.1 -lgcc )" ) > ./libgcc_s.so
  /nix/store/58nv799ffwvrcql2l4x7sqzb33d054pc-binutils-2.36/bin/ld: /build/build/./gcc/liblto_plugin.so: error loading plugin: /nix/store/xfd2dkxg0ha8z3a517ag2jvf6gwnplim-bootstrap-stage0-glibc/lib/libc.so.6: version `GLIBC_2.33' not found (required by /build/build/./gcc/liblto_plugin.so)
  collect2: error: ld returned 1 exit status
  make[3]: *** [Makefile:994: libgcc_s.so] Error 1
  make[3]: Leaving directory '/build/build/x86_64-unknown-linux-gnu/libgcc'
  make[2]: *** [Makefile:18294: all-stage1-target-libgcc] Error 2
  make[2]: Leaving directory '/build/build'
  make[1]: *** [Makefile:22554: stage1-bubble] Error 2
  make[1]: Leaving directory '/build/build'
  make: *** [Makefile:23445: profiledbootstrap] Error 2
error: --- Error --- nix-daemon
1 dependencies of derivation '/nix/store/s5igyqhv1khf7pvgzq70pxjqxpr4avg5-bootstrap-stage4-gcc-wrapper-10.2.0.drv' failed to build
error: --- Error --- nix-daemon
1 dependencies of derivation '/nix/store/sjc6ygmqh1jnrgky3a0snbz8fv3wdc9j-bash-4.4-p23.drv' failed to build
error: --- Error --- nix-daemon
1 dependencies of derivation '/nix/store/hmpk5in1v62wbv682kp31jai5c1ik1x5-bootstrap-stage4-stdenv-linux.drv' failed to build
error: --- Error --------------------------------------------------------------------------------------------------------- nix-build
1 dependencies of derivation '/nix/store/4mnw6mkkq3gcdiv4y4la7kq7s8i9r1x0-binutils-wrapper-2.36.drv' failed to build
```

It seems to expect `/nix/store/xfd2dkxg0ha8z3a517ag2jvf6gwnplim-bootstrap-stage0-glibc/lib/libc.so.6` to be glibc 2.33, but checking with `ldd` I see that it's 2.32-35:
```
❯ ldd /nix/store/xfd2dkxg0ha8z3a517ag2jvf6gwnplim-bootstrap-stage0-glibc/lib/libc.so.6
        /nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-glibc-2.27/lib/ld-linux-x86-64.so.2 => /nix/store/gafigwfaimlziam6qhw1m8dz4h952g1n-glibc-2.32-35/lib64/ld-linux-x86-64.so.2 (0x00007f18d165e000)
        linux-vdso.so.1 (0x00007fff03fc9000)
```

Not sure how to get around this

**Comment by lovesegfault** [comment]
I think this will need an update to our bootstrap binaries like https://github.com/NixOS/nixpkgs/pull/66528 did. I'll wait on advice on how to do this before proceeding.

**Comment by vcunat** [comment]
It seems to be 2.27.  Generally, I think updating tools like glibc or gcc very soon after its release often needs relatively lots of work (I'd probably prefer the patch approach #110570 for now).

Similarly, it would be nice if we could avoid updating the bootstrap tools often; I think I mostly described it [around here](https://github.com/NixOS/nixpkgs/pull/66528#issuecomment-582351637), but IIRC there were a few other less common platforms that also needed change, so altogether it would again cause quite some work.

**Comment by SuperSandro2000** [comment]
@lovesegfault can you fix the merge conflict?

**Comment by TredwellGit** [comment]
This also needs to be updated with the latest patches.

**Comment by vcunat** [comment]
So, I assume we'll keep the 21.05 release on glibc-2.32, right?

**Comment by Ma27** [comment]
I'd say no. At least I don't have time to push a glibc update forward. However I don't think it's such a big deal though since @TredwellGit is doing a pretty decent job with keeping the patchset of 2.32 up-to-date :blush:

**Comment by vcunat** [comment]
Yes, the stable branches in glibc repos seem nice, so it should be easy for us to cover all significant bugs (well, except for the expense of rebuilding everything).  Few upstream projects do this.

**Comment by TredwellGit** [comment]
I will post an updated version of this pull request later and will need someone to configure a Hydra jobset for it. Some programs require specific patching for 2.33.

**Comment by Ma27** [comment]
Yeah that's totally normal. May be helpful to look at the previous bumps (#95089, #79391, #66528), especially regarding the bootstrap tools stuff :)

**Comment by Ma27** [comment]
@lovesegfault @TredwellGit what are your current plans here? If I get a Hydra jobset, I'm happy to help / take over given that I already have some experience with breaking glibc updates :)

**Comment by lovesegfault** [comment]
> @lovesegfault @TredwellGit what are your current plans here? If I get a Hydra jobset, I'm happy to help / take over given that I already have some experience with breaking glibc updates :)
> 
> EDIT: Hmm, actually I could ask one of my colleagues with more permissions hydra.nixos.org :) But before I do that, I'd like to make sure I'm not duplicating any work.

I guess now that 21.05 was branched off we should kick this off again. I'm happy to let someone else take this over, I'm going to be very busy for the next couple of months and don't anticipate having bandwidth to take this on, unfortunately.

@Ma27 do you want me to close this PR and let you open a new one, or would you like to start pushing commits to this one?

**Comment by TredwellGit** [comment]
Please take over. I am not sure how to deal with https://github.com/NixOS/nixpkgs/pull/111616#issuecomment-771247040.

**Comment by Ma27** [comment]
> @Ma27 do you want me to close this PR and let you open a new one, or would you like to start pushing commits to this one?

I think that one's fine as it uses a branch from NixOS/nixpkgs :)

**Comment by lovesegfault** [comment]
> > @Ma27 do you want me to close this PR and let you open a new one, or would you like to start pushing commits to this one?
> 
> I think that one's fine as it uses a branch from NixOS/nixpkgs :)

Perfect, it's all yours!

**Comment by Ma27** [comment]
I'm not really sure if `config.inBootstrap` is really a good idea, but at least we have a POC :)

**Comment by Ma27** [comment]
Going to mark as a draft now. This isn't really reviewable (yet), so let's make that official :)

**Comment by Ma27** [comment]
Hydra jobset is at https://hydra.nixos.org/jobset/nixpkgs/glibc-2.33#tabs-evaluations btw.

**Comment by Ma27** [comment]
Temporarily rebasing this onto staging-next. I'll move the base branch back to `staging` when we're ready here, but this makes it easier to compare the branch against the `staging-next` jobset and probably has less problems :)

**Comment by Ma27** [comment]
According to https://hydra.nixos.org/eval/1674402?compare=staging-next there aren't many failures and the ones in the "newly failing builds" tab mostly don't seem related. Will start a new eval with the latest force-push and if that's done, I'd mark this ready for review and base this onto staging.

**Comment by TredwellGit** [comment]
The 2.33 patches need to be included:
https://github.com/NixOS/nixpkgs/blob/690496c4e545e68482b5c162a03f0a4f97d35373/pkgs/development/libraries/glibc/common.nix#L62-L70

**Comment by Ma27** [comment]
You're right, thanks for pointing that out!

**Comment by TredwellGit** [comment]
Also, you need to provide reproducible steps such that the bootstrap can be independently verified.

**Comment by Ma27** [comment]
> Also, you need to provide reproducible steps such that the bootstrap can be independently verified.

What's wrong with `nix-build pkgs/stdenv/linux/make-bootstrap-tools.nix`?

**Comment by TredwellGit** [comment]
On which commit?

**Comment by Ma27** [comment]
https://github.com/NixOS/nixpkgs/pull/111616/commits/224f1b428cf6d290a4c5901a2859e4717be60271 IIRC

**Comment by Ma27** [comment]
But please keep in mind that we'll have to rebuild these files anyways. I built the aarch64 ones on the nixos community builder and build artifacts from that machine shouldn't be trusted.

**Comment by Ma27** [comment]
@TredwellGit updated to `2.33-41` as described in the comment. Would you mind checking if this was correct? :) 
Also, are there any questions left re bootstrapping?

**Comment by TredwellGit** [comment]
The patches are correct.

**Comment by TredwellGit** [comment]
https://github.com/NixOS/nixpkgs/blob/df708b7746130af6c5e0b586addb43a71ac8acad/pkgs/stdenv/linux/bootstrap-files/x86_64.nix#L14-L22
```
$ git checkout 224f1b428cf6d290a4c5901a2859e4717be60271 && awk '{print $3}' "$(nix-build pkgs/top-level/release.nix -Q -A stdenvBootstrapTools.x86_64-linux.dist)/nix-support/hydra-build-products"
HEAD is now at 224f1b428cf stdenv bootstrap: upgrade x86_64-linux bootstrap tools & use temporary new files
/nix/store/f77gy7yg6rdfp5ckwyayzy776njpcwgv-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz
/nix/store/f77gy7yg6rdfp5ckwyayzy776njpcwgv-stdenv-bootstrap-tools/on-server/busybox

$ diff /nix/store/f77gy7yg6rdfp5ckwyayzy776njpcwgv-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz <(curl -s 'https://intranet.mayflower.de/s/XXTCb6MspFaJzPq/download/bootstrap-tools.tar.xz')
Binary files /nix/store/f77gy7yg6rdfp5ckwyayzy776njpcwgv-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz and /dev/fd/63 differ

$ diff /nix/store/f77gy7yg6rdfp5ckwyayzy776njpcwgv-stdenv-bootstrap-tools/on-server/busybox <(curl -s 'https://intranet.mayflower.de/s/XPNdMtjWTdiGGsG/download/busybox')
```

**Comment by Ma27** [comment]
The bootstrap tools you've built are already using the bootstrap-tools I added in the commit and I'm pretty sure that this makes a difference already. Apart from that I'm not sure which state exactly I used (it's not necessarily a commited state). In the end this was primarily to check whether the approach works fine.

How about building fresh bootstrap tools from the latest rev in the branch and check for their reproducibility then? Also we'd need @vcunat or @edolstra to push them to tarballs.nixos.org in the end, so at least another maintainer who can verify reproducibility.

**Comment by TredwellGit** [comment]
origin/release/2.33/master was updated: https://sourceware.org/git/?p=glibc.git;a=shortlog;h=refs/heads/release/2.33/master

**Comment by Ma27** [comment]
Rebased onto staging now and started a new evaluation. Since I didn't find any other broken packages (see https://hydra.nixos.org/eval/1673862) I think we're mostly done with the "fix fallout" phase.

The following things have to be done

* Review of the changes (especially my changes to the stdenv bootstrap!)
* Build a new bootstrap tarball, upload it to tarballs.nixos.org (and perhaps run a last Hydra eval)

cc @vcunat @FRidh @Ericson2314 it'd be very nice if you could take a look :)

**Comment by TredwellGit** [comment]
This needs to be updated for CVE-2021-33574.

**Comment by Ma27** [comment]
cc @Ericson2314 @FRidh @vcunat any chance to get a review here? :)

**Comment by lovesegfault** [comment]
@Ma27 Looks like this needs a rebase again :)

@Ericson2314, @FRidh, @vcunat This still needs a look from y'all :)

File: pkgs/build-support/cc-wrapper/default.nix [hunk 1]
```diff
@@ -27,6 +28,8 @@ assert !nativeTools ->
 assert !(nativeLibc && noLibc);
 assert (noLibc || nativeLibc) == (libc == null);
 
+let cc' = if config.inBootstrap or false && cc ? override then cc.override { enableLTO = false; } else cc; in
+let cc = cc'; in
```
Comment (SuperSandro2000) [inline pkgs/build-support/cc-wrapper/default.nix:32]: Maybe add a comment here?

**Comment by SuperSandro2000** [comment]
I can build some packages to see if something else broke if you resolve the merge conflict.

**Comment by TredwellGit** [comment]
Needs update to glibc-2.33-45-g58b90461ae.

**Comment by Ma27** [comment]
> I can build some packages to see if something else broke if you resolve the merge conflict.

That's very nice, but there's already a Hydra jobset in https://hydra.nixos.org/jobset/nixpkgs/glibc-2.33#tabs-evaluations (please don't be confused by the -11000 that happened after I based this from staging-next to staging and most of them are aborted builds). I haven't found any other issues, but in case you want to take a second look, that'd be helpful as well :) 

> Needs update to glibc-2.33-45-g58b90461ae.

I'm aware that the CVE we recently patched for 2.32 isn't fixed here and yes, I'm planning to do it before merging, however I'd prefer to get some reviews on the approach I've taken to fix stdenv bootstrap with this glibc so I know I'm on the right track in general :)

**Comment by Ma27** [comment]
ping @vcunat @FRidh @Ericson2314 any chance to gt a review? :)

File: pkgs/build-support/cc-wrapper/default.nix (hunk 1 above)
Comment (Ma27) [reply pkgs/build-support/cc-wrapper/default.nix:35]: done.

**Comment by Ma27** [comment]
@TredwellGit @lovesegfault rebased and updated glibc. I'd suggest we all build `pkgs/stdenv/linux/make-bootstrap-tools.nix` from 1e5eace3d9e759c3194cf98876499880f0960605 to verify them. As soon as @lovesegfault has access to tarballs.nixos.org (https://github.com/NixOS/nixos-org-configurations/pull/153), we can upload it there and update the bootstrap-file expressions.

I'd still like to have a review from e.g. @vcunat / @FRidh / @Ericson2314 after that, but then we should be good to go :)

**Comment by Ma27** [comment]
Also started a new evaluation for https://hydra.nixos.org/jobset/nixpkgs/glibc-2.33# .

**Comment by Ma27** [comment]
For x86_64-linux it is `/nix/store/q704cmhdlmcx9apmia16mbwh9pyig26v-stdenv-bootstrap-tools` and I got the following content-hashes:
```
[nix-shell:~/result/on-server]$ sha256sum  *
e06c0fc1dccc350c624608036d3e4015bab805727065a4169ecd45d013c41c60  bootstrap-tools.tar.xz
10ccc09257787de2813351cdf5448dcc8785912be96298b36431574eb67440b4  busybox
```

**Comment by vcunat** [comment]
This is suspicious:
```
$ nix build -f. glibc
$ ldd -r ./result/lib/libc.so.6
        /nix/store/mjnbpj0xwp47gc2c80kmmkd32gbkfr0y-glibc-2.33-45/lib/ld-linux-x86-64.so.2 => /nix/store/ikl21vjfq900ccbqg1xasp83kadw6q8y-glibc-2.32-46/lib64/ld-linux-x86-64.so.2 (0x00007fafd48b3000)
        linux-vdso.so.1 (0x00007ffd08ed9000)
undefined symbol: _dl_catch_error_ptr, version GLIBC_PRIVATE    (./result/lib/libc.so.6)
```
(the currently merged glibc shows no errors)

**Comment by Ma27** [comment]
> EDIT: it's most likely an issue with ldd when mixing libc versions.

Yep, that's what I just observed as well, when running `nix-shell` on this branch and doing the `ldd` then, it's perfectly fine. Not sure what to do about this, though.

FWIW the symbol seems to have been added in https://sourceware.org/git/?p=glibc.git;a=commitdiff;h=f4cba6ca1e5ff2c229af6d474c3ef2c0df2fae52

**Comment by vcunat** [comment]
This ldd issue seems unrelated to the version update, really.  I don't know if it could be fixed.

**Comment by vcunat** [comment]
I wanted to avoid repeating the gcc LTO problem on future glibc updates.  I found a hack that seems to work well and it's probably even simpler than updating all the bootstrap tools (every time), so I'd personally go that way instead.  What do you think?  https://github.com/vcunat/nixpkgs/commits/p/glibc-2.33

(I tested building stdenv on x86_64 and aarch64.)

**Comment by vcunat** [comment]
- I suspect it depends on which symbols get added by glibc.
- Well, noone counts on a linker plugin being loaded against a different libc than the `ld` loading that plugin.  I'd say that's specific to Nix*, really.  Glibc should be quite resilient on upgrades but not on downgrades; `ld` is first and thus sets the libc, so the plugin experiences a libc downgrade.
- I'm not aware of such a possibility (and see up/down above), and the whole bootstrap succeeded for me.  The final stdenv or anything later shouldn't be affected by the hack.

**Comment by Ma27** [comment]
Thanks a lot for the clarification!

>  The final stdenv or anything later shouldn't be affected by the hack.

Agreed, that's in the end the assumption I based my original work on. But in the end I agree that your approach is more desirable.

So feel free to push your changes to this branch. We may want to start another jobset eval on Hydra, but then it should be good to go from my PoV :)

**Comment by vcunat** [comment]
OK, re-tested and pushed.

I expect noone else will react anymore... anyway, I kept the changes fast-forward for now, but before merging it will be better to cancel out the commit-revert pairs and perhaps squash the version updates (so that history contains fewer of those `*.patch.gz` files).

**Comment by Ma27** [comment]
Would be OK for me. Do you want to take care of this or should I do it? :) 

Also, thanks a lot for your work here! :)

**Comment by vcunat** [comment]
I don't care who does.

In any case, now I expect we wait for Hydra for a bit, though my changes shouldn't affect anything but the bootstrap itself, and [previous](https://hydra.nixos.org/eval/1674402) failure counts certainly seemed low enough for merge to staging... so I don't think we'll need to wait for majority to finish.

**Comment by vcunat** [comment]
I cleaned up the history, zero diff for the force-push.

**Comment by vcunat** [comment]
The Hydra comparison is ruined by an unrelated problem in `python3Packages.faker`, but anyway if any regressions from the glibc update appear, it should be possible for us to handle during the usual `staging-next` stabilization.  Bootstrapping certainly was tested there for the three `-linux` platforms.

**Comment by lovesegfault** [comment]
Thanks everyone for all the work on this!

For future reference, I now have access to `nixpkgs-tarballs`, so when the time comes for another bump and we _do_ end up needing to put something there please don't hesitate to ping.

**Comment by vcunat** [comment]
TODO:
 - [ ] zig conflicts https://hydra.nixos.org/build/147355229
 
 (there are way more serious issues on staging-next ATM)

**Comment by dotlambda** [comment]
The PROJ maintainers should have been pinged.

**Comment by vcunat** [comment]
🤔 I thought that happens automatically, but perhaps there were too many packages affected here or something.

**Comment by lovesegfault** [comment]
My fault, sorry @dotlambda!

Added a note to my workflow to manually ping in the future.

**Pull request merged**